File: packages/ui/src/components/ui/finance/wallets/wallet-images.ts
/**
 * Wallet image definitions for bank and mobile payment providers.
 * Images are stored in /media/apps/tutrack/ with variants:
 * - {id}-square.svg (used for icons)
 * - {id}-horizontal-fit.svg
 * - {id}-in-a-frame.svg
 */

export interface WalletImage {
  /** Unique identifier (e.g., "bidv", "momo") */
  id: string;
  /** Display name */
  name: string;
  /** Category: bank or mobile */
  category: 'bank' | 'mobile';
}

/**
 * Bank images (64 banks)
 */
export const WALLET_BANK_IMAGES: WalletImage[] = [
  { id: 'abbank', name: 'ABBank', category: 'bank' },
  { id: 'acb', name: 'ACB', category: 'bank' },
  { id: 'anz', name: 'ANZ', category: 'bank' },
  { id: 'argibank', name: 'Agribank', category: 'bank' },
  { id: 'bac-a-bank', name: 'Bac A Bank', category: 'bank' },
  { id: 'bangkok-bank', name: 'Bangkok Bank', category: 'bank' },
  { id: 'bank-of-china', name: 'Bank of China', category: 'bank' },
  { id: 'baoviet-bank', name: 'BaoViet Bank', category: 'bank' },
  { id: 'bidc', name: 'BIDC', category: 'bank' },
  { id: 'bidv', name: 'BIDV', category: 'bank' },
  { id: 'cake-by-vpbank', name: 'Cake by VPBank', category: 'bank' },
  { id: 'cathay-bank', name: 'Cathay Bank', category: 'bank' },
  { id: 'cb', name: 'CB', category: 'bank' },
  { id: 'chase-bank', name: 'Chase Bank', category: 'bank' },
  { id: 'cimb', name: 'CIMB', category: 'bank' },
  { id: 'citibank', name: 'Citibank', category: 'bank' },
  { id: 'commonwealth-bank', name: 'Commonwealth Bank', category: 'bank' },
  { id: 'deutsche-bank', name: 'Deutsche Bank', category: 'bank' },
  { id: 'donga-bank', name: 'DongA Bank', category: 'bank' },
  { id: 'dsb', name: 'DSB', category: 'bank' },
  { id: 'eximbank', name: 'Eximbank', category: 'bank' },
  { id: 'gpbank', name: 'GPBank', category: 'bank' },
  { id: 'hdbank', name: 'HDBank', category: 'bank' },
  { id: 'hong-leong-bank', name: 'Hong Leong Bank', category: 'bank' },
  { id: 'hsbc', name: 'HSBC', category: 'bank' },
  { id: 'industrial-bank-of-korea', name: 'Industrial Bank of Korea', category: 'bank' },
  { id: 'ivb', name: 'IVB', category: 'bank' },
  { id: 'kb', name: 'KB', category: 'bank' },
  { id: 'keb-hana-bank', name: 'KEB Hana Bank', category: 'bank' },
  { id: 'kienlong-bank', name: 'Kienlong Bank', category: 'bank' },
  { id: 'lienviet-postbank', name: 'LienViet PostBank', category: 'bank' },
  { id: 'maybank', name: 'Maybank', category: 'bank' },
  { id: 'mb-bank', name: 'MB Bank', category: 'bank' },
  { id: 'mega-international-commercial-bank', name: 'Mega Bank', category: 'bank' },
  { id: 'mizuho', name: 'Mizuho', category: 'bank' },
  { id: 'msb', name: 'MSB', category: 'bank' },
  { id: 'nam-a-bank', name: 'Nam A Bank', category: 'bank' },
  { id: 'napas', name: 'NAPAS', category: 'bank' },
  { id: 'ncb', name: 'NCB', category: 'bank' },
  { id: 'ocb', name: 'OCB', category: 'bank' },
  { id: 'ocean-bank', name: 'Ocean Bank', category: 'bank' },
  { id: 'pg-bank', name: 'PG Bank', category: 'bank' },
  { id: 'public-bank', name: 'Public Bank', category: 'bank' },
  { id: 'pvcom-bank', name: 'PVcomBank', category: 'bank' },
  { id: 'scb', name: 'SCB', category: 'bank' },
  { id: 'scotiabank', name: 'Scotiabank', category: 'bank' },
  { id: 'sea-bank', name: 'SeABank', category: 'bank' },
  { id: 'shb', name: 'SHB', category: 'bank' },
  { id: 'shinhan-bank', name: 'Shinhan Bank', category: 'bank' },
  { id: 'smfg', name: 'SMFG', category: 'bank' },
  { id: 'standard-chartered', name: 'Standard Chartered', category: 'bank' },
  { id: 'techcombank', name: 'Techcombank', category: 'bank' },
  { id: 'timo', name: 'Timo', category: 'bank' },
  { id: 'tpbank', name: 'TPBank', category: 'bank' },
  { id: 'tyme', name: 'Tyme', category: 'bank' },
  { id: 'tymebank', name: 'TymeBank', category: 'bank' },
  { id: 'ubsp', name: 'UBSP', category: 'bank' },
  { id: 'vib', name: 'VIB', category: 'bank' },
  { id: 'viet-a-bank', name: 'VietABank', category: 'bank' },
  { id: 'viet-capital-bank', name: 'Viet Capital Bank', category: 'bank' },
  { id: 'vietcombank', name: 'Vietcombank', category: 'bank' },
  { id: 'vietinbank', name: 'VietinBank', category: 'bank' },
  { id: 'vpbank', name: 'VPBank', category: 'bank' },
  { id: 'woori-bank', name: 'Woori Bank', category: 'bank' },
];

/**
 * Mobile payment images (20 providers)
 */
export const WALLET_MOBILE_IMAGES: WalletImage[] = [
  { id: 'apota', name: 'Apota', category: 'mobile' },
  { id: 'baokim', name: 'BaoKim', category: 'mobile' },
  { id: 'ecpay', name: 'ECPay', category: 'mobile' },
  { id: 'grabpay-by-moca', name: 'GrabPay by Moca', category: 'mobile' },
  { id: 'moca', name: 'Moca', category: 'mobile' },
  { id: 'momo', name: 'MoMo', category: 'mobile' },
  { id: 'nextpay', name: 'NextPay', category: 'mobile' },
  { id: 'ngan-luong', name: 'Ngan Luong', category: 'mobile' },
  { id: 'payme', name: 'PayMe', category: 'mobile' },
  { id: 'payoo', name: 'Payoo', category: 'mobile' },
  { id: 'shopeepay', name: 'ShopeePay', category: 'mobile' },
  { id: 'truemoney', name: 'TrueMoney', category: 'mobile' },
  { id: 'viettel-money', name: 'Viettel Money', category: 'mobile' },
  { id: 'viettel-pay', name: 'ViettelPay', category: 'mobile' },
  { id: 'vimo', name: 'Vimo', category: 'mobile' },
  { id: 'vinid', name: 'VinID', category: 'mobile' },
  { id: 'vnpay', name: 'VNPay', category: 'mobile' },
  { id: 'vnpay-money', name: 'VNPay Money', category: 'mobile' },
  { id: 'vtcpay', name: 'VTCPay', category: 'mobile' },
  { id: 'zalopay', name: 'ZaloPay', category: 'mobile' },
];

/**
 * All wallet images combined
 */
export const ALL_WALLET_IMAGES: WalletImage[] = [
  ...WALLET_BANK_IMAGES,
  ...WALLET_MOBILE_IMAGES,
];

/**
 * Get the full path to a wallet image
 * @param imageSrc The image source identifier (e.g., "bank/bidv" or "mobile/momo")
 * @returns The full path to the square SVG image
 */
export function getWalletImagePath(imageSrc: string): string {
  return `/media/apps/tutrack/${imageSrc}-square.svg`;
}

/**
 * Get wallet image metadata by its full image_src value
 * @param imageSrc The image source (e.g., "bank/bidv")
 * @returns The WalletImage metadata or undefined if not found
 */
export function getWalletImageById(imageSrc: string): WalletImage | undefined {
  const [category, id] = imageSrc.split('/');
  if (category === 'bank') {
    return WALLET_BANK_IMAGES.find((img) => img.id === id);
  }
  if (category === 'mobile') {
    return WALLET_MOBILE_IMAGES.find((img) => img.id === id);
  }
  return undefined;
}

/**
 * Build the image_src value from category and id
 * @param category The category ('bank' or 'mobile')
 * @param id The image id
 * @returns The combined image_src value (e.g., "bank/bidv")
 */
export function buildImageSrc(category: 'bank' | 'mobile', id: string): string {
  return `${category}/${id}`;
}

/**
 * Filter wallet images by search query
 * Matches against name and id
 */
export function filterWalletImages(
  images: WalletImage[],
  query: string
): WalletImage[] {
  if (!query.trim()) return images;
  const searchTerm = query.toLowerCase().trim();
  return images.filter(
    (img) =>
      img.name.toLowerCase().includes(searchTerm) ||
      img.id.toLowerCase().includes(searchTerm)
  );
}

File: apps/mobile/lib/features/finance/utils/wallet_images.dart
import 'package:mobile/core/config/api_config.dart';

enum WalletImageCategory { bank, mobile }

class WalletImageOption {
  const WalletImageOption({
    required this.id,
    required this.name,
    required this.category,
  });

  final String id;
  final String name;
  final WalletImageCategory category;

  String get imageSrc => '${category.name}/$id';
}

const walletBankImages = <WalletImageOption>[
  WalletImageOption(
    id: 'abbank',
    name: 'ABBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'acb', name: 'ACB', category: WalletImageCategory.bank),
  WalletImageOption(id: 'anz', name: 'ANZ', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'argibank',
    name: 'Agribank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'bac-a-bank',
    name: 'Bac A Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'bangkok-bank',
    name: 'Bangkok Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'bank-of-china',
    name: 'Bank of China',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'baoviet-bank',
    name: 'BaoViet Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'bidc',
    name: 'BIDC',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'bidv',
    name: 'BIDV',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'cake-by-vpbank',
    name: 'Cake by VPBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'cathay-bank',
    name: 'Cathay Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'cb', name: 'CB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'chase-bank',
    name: 'Chase Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'cimb',
    name: 'CIMB',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'citibank',
    name: 'Citibank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'commonwealth-bank',
    name: 'Commonwealth Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'deutsche-bank',
    name: 'Deutsche Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'donga-bank',
    name: 'DongA Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'dsb', name: 'DSB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'eximbank',
    name: 'Eximbank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'gpbank',
    name: 'GPBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'hdbank',
    name: 'HDBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'hong-leong-bank',
    name: 'Hong Leong Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'hsbc',
    name: 'HSBC',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'industrial-bank-of-korea',
    name: 'Industrial Bank of Korea',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'ivb', name: 'IVB', category: WalletImageCategory.bank),
  WalletImageOption(id: 'kb', name: 'KB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'keb-hana-bank',
    name: 'KEB Hana Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'kienlong-bank',
    name: 'Kienlong Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'lienviet-postbank',
    name: 'LienViet PostBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'maybank',
    name: 'Maybank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'mb-bank',
    name: 'MB Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'mega-international-commercial-bank',
    name: 'Mega Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'mizuho',
    name: 'Mizuho',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'msb', name: 'MSB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'nam-a-bank',
    name: 'Nam A Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'napas',
    name: 'NAPAS',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'ncb', name: 'NCB', category: WalletImageCategory.bank),
  WalletImageOption(id: 'ocb', name: 'OCB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'ocean-bank',
    name: 'Ocean Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'pg-bank',
    name: 'PG Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'public-bank',
    name: 'Public Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'pvcom-bank',
    name: 'PVcomBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'scb', name: 'SCB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'scotiabank',
    name: 'Scotiabank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'sea-bank',
    name: 'SeABank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'shb', name: 'SHB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'shinhan-bank',
    name: 'Shinhan Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'smfg',
    name: 'SMFG',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'standard-chartered',
    name: 'Standard Chartered',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'techcombank',
    name: 'Techcombank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'timo',
    name: 'Timo',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'tpbank',
    name: 'TPBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'tyme',
    name: 'Tyme',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'tymebank',
    name: 'TymeBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'ubsp',
    name: 'UBSP',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(id: 'vib', name: 'VIB', category: WalletImageCategory.bank),
  WalletImageOption(
    id: 'viet-a-bank',
    name: 'VietABank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'viet-capital-bank',
    name: 'Viet Capital Bank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'vietcombank',
    name: 'Vietcombank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'vietinbank',
    name: 'VietinBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'vpbank',
    name: 'VPBank',
    category: WalletImageCategory.bank,
  ),
  WalletImageOption(
    id: 'woori-bank',
    name: 'Woori Bank',
    category: WalletImageCategory.bank,
  ),
];

const walletMobileImages = <WalletImageOption>[
  WalletImageOption(
    id: 'apota',
    name: 'Apota',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'baokim',
    name: 'BaoKim',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'ecpay',
    name: 'ECPay',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'grabpay-by-moca',
    name: 'GrabPay by Moca',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'moca',
    name: 'Moca',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'momo',
    name: 'MoMo',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'nextpay',
    name: 'NextPay',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'ngan-luong',
    name: 'Ngan Luong',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'payme',
    name: 'PayMe',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'payoo',
    name: 'Payoo',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'shopeepay',
    name: 'ShopeePay',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'truemoney',
    name: 'TrueMoney',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'viettel-money',
    name: 'Viettel Money',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'viettel-pay',
    name: 'ViettelPay',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'vimo',
    name: 'Vimo',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'vinid',
    name: 'VinID',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'vnpay',
    name: 'VNPay',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'vnpay-money',
    name: 'VNPay Money',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'vtcpay',
    name: 'VTCPay',
    category: WalletImageCategory.mobile,
  ),
  WalletImageOption(
    id: 'zalopay',
    name: 'ZaloPay',
    category: WalletImageCategory.mobile,
  ),
];

const walletImageOptions = <WalletImageOption>[
  ...walletBankImages,
  ...walletMobileImages,
];

String walletImageUrl(String imageSrc) {
  final baseUrl = ApiConfig.baseUrl.replaceFirst(RegExp(r'/$'), '');
  return '$baseUrl/media/apps/tutrack/$imageSrc-square.svg';
}

WalletImageOption? findWalletImageBySrc(String? imageSrc) {
  if (imageSrc == null || imageSrc.trim().isEmpty) {
    return null;
  }
  for (final item in walletImageOptions) {
    if (item.imageSrc == imageSrc) {
      return item;
    }
  }
  return null;
}

List<WalletImageOption> filterWalletImages(
  List<WalletImageOption> source,
  String query,
) {
  final normalized = query.trim().toLowerCase();
  if (normalized.isEmpty) {
    return source;
  }

  return source
      .where(
        (item) =>
            item.name.toLowerCase().contains(normalized) ||
            item.id.toLowerCase().contains(normalized),
      )
      .toList(growable: false);
}
